feat: Add Programmed External Keys.

TheShubham99 opened the pull request:
This example shows how to use external buttons/keys as a keyboard using Keyboard Library

TheShubham99 committed:
Create Programmed External Keys.ino

per1234 commented on examples/Programmable Buttons/Programmed External Keys.ino:
```diff
@@ -0,0 +1,68 @@
+      Keyboard.releaseAll();
+    }
+}
+void loop() {
```
```suggestion

void loop() {
```
Be consistent in your use of blank lines.

per1234 commented on examples/Programmable Buttons/Programmed External Keys.ino:
```diff
@@ -0,0 +1,68 @@
+      delay(100);
+      Keyboard.releaseAll();
+    }
+}
```
What happened to the other two buttons?

TheShubham99 replied:
yes, they are programmable but I haven't added the functionality for them. should I mention it in the comments or should I remove those keys?

per1234 replied:
If you're not going to use them, remove them, since they will just cause confusion for beginners. If someone wants to expand on the sketch, they should be able to do that easily enough.

I see you started on that by changing `pincount` to 2, but you didn't remove the unused pins from `btnPin[]`, `prevBtnState[]`, `lastDebounceTime[]`. You didn't change the index numbers in `outputAction()`, which means the sketch doesn't even work now (because `currentButton` can never be 2). I guess you forgot the thing I said about thoroughly testing your work.

TheShubham99 replied:
Done.

per1234 commented on examples/Programmable Buttons/Programmed External Keys.ino:
```diff
@@ -0,0 +1,68 @@
+}
+
+// Place the actions needed after pin press here
+int outputAction(int currentButton) {
```
```suggestion
void outputAction(int currentButton) {
```
The function doesn't return anything, so the appropriate type is `void`.

per1234 commented on examples/Programmable Buttons/Programmed External Keys.ino:
```diff
@@ -0,0 +1,68 @@
+int btnState[] = {0, 0, 0, 0};
+int prevbtnState[] = {HIGH, HIGH, HIGH, HIGH};
+
+long lastDebounceTime[] = {0, 0, 0, 0};
```
```suggestion
unsigned long lastDebounceTime[] = {0, 0, 0, 0};
```
`millis()` returns `unsigned long`, so the appropriate type for this variable is `unsigned long`

per1234 commented on examples/Programmable Buttons/Programmed External Keys.ino:
```diff
@@ -0,0 +1,68 @@
+int prevbtnState[] = {HIGH, HIGH, HIGH, HIGH};
+
+long lastDebounceTime[] = {0, 0, 0, 0};
+long debounceDelay = 50;
```
```suggestion
unsigned long debounceDelay = 50;
```
Use of signed type results in the compiler warning:
```
warning: comparison between signed and unsigned integer expressions
```

per1234 commented on examples/Programmable Buttons/Programmed External Keys.ino:
```diff
@@ -0,0 +1,68 @@
+
+    prevbtnState[thisPin] = btnState[thisPin];
+  }
+}
```
```suggestion
}

```
End files in a newline to avoid meaningless noise in the diff during future modifications to the file.

TheShubham99 replied:
Ok. Thanks for the review :)

per1234 reviewed (changes requested):
Arduino sketch filenames may not contain spaces.

The sketch filename must exactly match the folder name:
```
Keyboard
|_examples
   |_ProgrammableButtons
      |_ProgrammableButtons.ino
```

Please do a **Tools > Auto Format** if using the Arduino IDE or a **Ctrl + B** if using Arduino Web Editor to ensure the formatting of your code matches that of the other official Arduino example sketches.

Please always thoroughly test your code before submitting contributions.

Please set **File > Preferences > Compiler Warnings** to "All" and make sure your code doesn't generate any warnings.

TheShubham99 committed:
Apply suggestions from code review

Formating Done

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 commented:
I will take care of all other suggestions too.

TheShubham99 committed:
Update examples/Programmable Buttons/Programmed External Keys.ino

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 committed:
Update examples/Programmable Buttons/Programmed External Keys.ino

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 committed:
Update examples/Programmable Buttons/Programmed External Keys.ino

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 committed:
Update examples/Programmable Buttons/Programmed External Keys.ino

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 committed:
Update examples/Programmable Buttons/Programmed External Keys.ino

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 committed:
Apply suggestions from code review

Control structures simplified

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 committed:
Update examples/Programmable Buttons/Programmed External Keys.ino

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 committed:
Update examples/Programmable Buttons/Programmed External Keys.ino

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 committed:
Update examples/Programmable Buttons/Programmed External Keys.ino

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 commented on examples/Programmable Buttons/Programmed External Keys.ino:
```diff
@@ -0,0 +1,68 @@
+#include "Keyboard.h"
+
+const int btnPin[] = {2, 3, 4, 5};
+int pincount = 4;
```
I will take care of that typo too. Thank you :)

per1234 replied:
@TheShubham99 this one still isn't resolved.

TheShubham99 committed:
Code Readability & Optimization

TheShubham99 commented on examples/Programmable Buttons/Programmed External Keys.ino:
```diff
@@ -0,0 +1,68 @@
+const int btnPin[] = {2, 3, 4, 5};
+int pincount = 4;
+int btnState[] = {0, 0, 0, 0};
+int prevbtnState[] = {HIGH, HIGH, HIGH, HIGH};
```
Sure. I will take care of that too.

TheShubham99 committed:
Updating ProgrammableButtons

Example for native USB Arduino boards (e.g., Leonardo, Micro, MKR, Nano 33 IoT, Zero, Due)

- When the first button is pressed, the Ctrl + C keyboard shortcut (copy) is emulated.

-When the second button is pressed, the Ctrl + V keyboard shortcut (paste) is emulated.

TheShubham99 reviewed (commented):
@per1234 I think I have done all the suggested changes. 
Let me know if there are still any issues/suggestions, which need to be resolved. 
**Thank you for your time & suggestions :)**

TheShubham99 committed:
Delete Programmed External Keys.ino

per1234 commented on examples/Programmable Buttons/Programmed External Keys.ino:
```diff
@@ -0,0 +1,68 @@
+
+#include "Keyboard.h"
+
+const int btnPin[] = {2, 3, 4, 5};
```
@TheShubham99 this one still isn't resolved. Because of this, your sketch currently doesn't compile:
```
ProgrammableButtons:21:22: error: array bound is not an integer constant before ']' token

 int btnState[pincount];

                      ^
```

TheShubham99 replied:
Resolved

per1234 commented on examples/ProgrammableButtons/ProgrammableButtons.ino:
```diff
@@ -0,0 +1,63 @@
+  for (int thisPin = pincount - 1; thisPin >= 0; thisPin--) {
+    btnState[thisPin] = digitalRead(btnPin[thisPin]);
+
+    if ((btnState[thisPin] != prevBtnState[thisPin]) && (btnState[thisPin] == HIGH)) {
```
```suggestion
    if ((btnState[thisPin] != prevBtnState[thisPin]) && (btnState[thisPin] == LOW)) {
```
Although the previous code does work, it triggers when the button is released. It is standard for a trigger to occur when the button is pushed, not released. Thus `LOW` is the better choice in this line.

TheShubham99 replied:
I got your point and Thank you very much for the suggestion. :)

But I wanted to ask that as this example is for beginners & in all Arduino examples, the trigger is on when key Reading is HIGH.

Also, the following instructions are from the Arduino official page, where all the beginners will learn.
> When the pushbutton is open (unpressed) there is no connection between the two legs of the pushbutton, so the pin is connected to ground (through the pull-down resistor) and we read a LOW. When the button is closed (pressed), it makes a connection between its two legs, connecting the pin to 5 volts, so that we read a HIGH.

[Link](https://www.arduino.cc/en/tutorial/button)

That is why I think we should keep this statement as it is.
```suggestion
    if ((btnState[thisPin] != prevBtnState[thisPin]) && (btnState[thisPin] == HIGH)) {
```

Please let me know if it is ok/or not?

per1234 replied:
>in all Arduino examples, the trigger is on when key Reading is HIGH.

Wrong. That is correct when you have a button connected between the IO pin and Vcc, with an external pull-down resistor as in the [Button tutorial](https://www.arduino.cc/en/Tutorial/Button). I guess they thought it would be more clear for beginners to have an external resistor and the HIGH == pressed logic.

However, there are plenty of examples that use the opposite logic. That's exactly the circuit you specified in the comment and the code. You specified that the button should be connected between the pin and ground, and you set the internal pull-up resistor (in the original code by using `digitalWrite` to set the pin `HIGH` after setting the pin mode to `INPUT`). In this style of circuit, the pin will be LOW when the button is pressed, as explained in this tutorial:
https://www.arduino.cc/en/Tutorial/InputPullupSerial

For any more advanced example like this one, it's best to use the internal pull-up resistor approach, as you have, because that circuit is easier to build.

TheShubham99 replied:
@per1234 **Thank you for clearing my doubt.

Made the changes. Please review it again :)**

TheShubham99 committed:
Update examples/ProgrammableButtons/ProgrammableButtons.ino

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 committed:
Update ProgrammableButtons.ino

TheShubham99 committed:
Update ProgrammableButtons.ino

TheShubham99 committed:
Update ProgrammableButtons.ino

Making Trigger when pinstate is HIGH.

TheShubham99 reviewed (commented):
@per1234 
 Thank you for guiding me through this :) 👍 

> **Please check this https://github.com/arduino-libraries/Keyboard/pull/33#discussion_r394915672 comment too.**

I have done all the requested changes. Please review..

TheShubham99 committed:
Update ProgrammableButtons.ino

TheShubham99 reviewed (commented):
@per1234 
Done with all the requested changes. (including the button trigger)

per1234 commented on examples/ProgrammableButtons/ProgrammableButtons.ino:
```diff
@@ -0,0 +1,64 @@
+int prevBtnState[] = {HIGH, HIGH}; 
+
+unsigned long lastDebounceTime[] = {0, 0};
+unsigned long debounceDelay = 50;
```
```suggestion
const unsigned long debounceDelay = 50;
```
Be consistent with the use of `const`. If you're going to use it one place in the code, use it everywhere.

per1234 commented on examples/ProgrammableButtons/ProgrammableButtons.ino:
```diff
@@ -0,0 +1,64 @@
+}
+
+// Place the actions needed after pin press here
+void outputAction(int currentButton) {
```
```suggestion
void outputAction(const int currentButton) {
```
Be consistent with the use of `const`. If you're going to use it one place in the code, use it everywhere.

TheShubham99 committed:
Update examples/ProgrammableButtons/ProgrammableButtons.ino

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 committed:
Update examples/ProgrammableButtons/ProgrammableButtons.ino

Co-Authored-By: per1234 <accounts@perglass.com>

TheShubham99 reviewed (commented):
**Thanks @per1234  👍  :)
I have implemented all of the required changes. :)** 

Please review it again :)

TheShubham99 commented:
@per1234  Can we merge the Example?

per1234 commented:
@TheShubham99 I'm not the person who decides which pull requests get merged. I only share my opinions in PR reviews in hopes it will be helpful.

Arduino's developers have a huge amount of software to maintain and are always working on very important and exciting projects. When you submit a pull request, you need to be willing to be patient. We should avoid making unproductive comments on the pull request threads, as the "noise" may cause the developers to be less productive in their important work, and thus result in longer delays before they can find the time to review non-essential pull requests.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/arduino-libraries/Keyboard?pullRequest=33) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/arduino-libraries/Keyboard?pullRequest=33) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/arduino-libraries/Keyboard?pullRequest=33) it.</sub>
